exercises/diffie-hellman/tests/diffie-hellman.rs

```diff
@@ -0,0 +1,60 @@
+extern crate diffie_hellman;
+
+use diffie_hellman::*;
+
+#[test]
+fn test_private_key_in_range_key() {
+    let primes: Vec<u64> = vec![5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 773, 967,
+                                3461, 6131];
+    let private_keys: Vec<u64> = primes.iter().map(|x| private_key(*x)).collect();
+
+    for i in 0..primes.len() {
+        assert!(1 <= private_keys[i] && private_keys[i] < primes[i]);
+    }
+}
+
+#[test]
+#[ignore]
+fn test_public_key_correct() {
+    let p: u64 = 23;
+    let g: u64 = 5;
+
+    let private_key: u64 = 6;
+    let expected: u64 = 8;
+
+    assert_eq!(public_key(p, g, private_key), expected);
+}
+
+
+#[test]
+#[ignore]
+fn test_secret_key_correct() {
+    let p: u64 = 11;
+    let g: u64 = 7;
+
+    let private_key = 7;
+    let public_key = public_key(p, g, private_key);
+    let secret = secret(p, public_key, private_key);
```

Comment:
> > so, to confirm, this is showing what happens when someone effectively performs a DH key exchange with themselves, right? Since they're using their own public key and their own private key.
> 
> In the last test `test_changed_secret_key()` they exchange their public keys.

I agree and now I am wanting to talk about `test_secret_key_correct`.

It appears the purpose of it is to make sure the `secret` function is correct. This is a good purpose; there absolutely should be a test that confirms this.

There is an interesting thing about the inputs to this function. We're asking for (the secret that results from (my own
 private key) and (the public key corresponding to my own private key)), which means I have performed a DH key exchange with myself. Again, now I show my ignorance, *maybe* this is a reasonable thing to do, because supposedly now I have a secret that I and only I know (it's `g ** (a ** 2) mod p`), I just don't know about it. If it is unusual to DH key exchange with myself, then I would not want the tests to show it as an example, since it would give people the wrong idea.

So I was thinking of having one of four things happen:

* Show me something that tells that it makes sense to DH key exchange with myself
* That I was actually mistaken, this is not a DH key exchange with myself
* Use a different (but still fixed!) private key in the call to `public_key` instead of the same one used in `secret`
* Use a fixed value for `public_key` instead of calling the `public_key` function here (in this case `g` can be deleted)

Reply:
```That I was actually mistaken, this is not a DH key exchange with myself```

There is a public key exchange between Alice(A) and Bob(B).
After the public key exchange, they both can generate a **secret session key** from their **own private key** and the **opposite public key**.
Alice can then verify that the message is from Bob because he has encrypted the message with the secret session key which only they both share.

Reply:
> **their own private key** and **the opposite public key**

I do not believe this is an opposite public key. Because the same `private_key` is used in `let public_key = public_key(p, g, private_key)` and `let secret = secret(p, public_key, private_key)`, The public key was derived from the same private key. For it to be an opposite public key, there must be `let public_key = public_key(p, g, some_other_private_key)`

Reply:
In anticipation of a possible response "it just so happens that by random chance, Alice and Bob have the same private key", I'll accept that by adding a fifth option:

* Add a line that says something like `let other_private_key = private_key; // by some coincidence, it just so happens that the two parties have the same private key` to make it clear that there is in fact a second party to this exchange, but they just happen to have the same key and it is not by design that they always have the same key.

Reply:
EBWODP: with the fifth option, then there would also be `let public_key = public_key(p, g, other_private_key)` of course.

Reply:
Ah now I get where the problem is. I thought you would still talk about the ```test_changed_secret_key()``` but you were talking about the ```test_secret_key_correct()```.
Yes, you are right there should be a public key from b and a private key from a for example.